Add EGL Integration

TheNeikos opened the pull request:
This PR adds vertex buffers to surfaces and prepares the egl drawing to the monitor.
- [x] Get the monitor to display EGL things
- [x] Add the vertex buffer to surfaces
- [x] Add a global redraw loop
- [x] Make it work

TheNeikos commented:
The make it work part is reaally hard. EGL is currently segfault deep into the library, and I don't why /yet/.

neithernut commented:
Looks like you have to rebase

TheNeikos commented:
Definitely

TheNeikos commented:
Made it work tho

so I got that going for me

matthiasbeyer commented on src/compositor/buffer/gbm.h:
```diff
@@ -37,50 +37,65 @@
  * Variable which holds the type information about the ws_frame_buffer type
  */
-extern ws_buffer_type_id WS_OBJECT_TYPE_ID_GBM_BUFFER;
+extern ws_object_type_id WS_OBJECT_TYPE_ID_gbm_surface;
```
Search&replace failure here.

matthiasbeyer commented:
You renamed the `gbm_buffer` to `gbm_surface` in this PR. Would it make sense to move the appropriate file out of the `buffer/` subdir, then?

matthiasbeyer commented on src/compositor/buffer/gbm.c:
```diff
@@ -54,67 +65,51 @@ gbm_buffer_deinit(
-    tmp->obj.obj.obj.id = (ws_object_type_id *) &WS_OBJECT_TYPE_ID_GBM_BUFFER;
+    ws_object_init(&tmp->obj);
+    tmp->obj.settings |= WS_OBJECT_HEAPALLOCED;
+    tmp->obj.id = (ws_object_type_id *) &WS_OBJECT_TYPE_ID_GBM_SURFACE;
```
Do you actually still need the cast here?

matthiasbeyer commented on src/compositor/buffer/gbm.c:
```diff
@@ -54,67 +65,51 @@ gbm_buffer_deinit(
-ws_buffer_type_id WS_OBJECT_TYPE_ID_GBM_BUFFER = {
-    .type = {
+ws_object_type_id WS_OBJECT_TYPE_ID_GBM_SURFACE = {
         .supertype  = (ws_object_type_id*) &WS_OBJECT_TYPE_ID_RAW_BUFFER,
```
This is not right anymore. The supertype is `ws_object` now, actually.

matthiasbeyer commented on src/compositor/buffer/gbm.c:
```diff
@@ -54,67 +65,51 @@ gbm_buffer_deinit(
-    .type = {
+ws_object_type_id WS_OBJECT_TYPE_ID_GBM_SURFACE = {
         .supertype  = (ws_object_type_id*) &WS_OBJECT_TYPE_ID_RAW_BUFFER,
         .typestr    = "ws_frame_buffer",
```
Should be fixed as well.

matthiasbeyer commented on src/compositor/buffer/gbm.c:
```diff
@@ -122,6 +117,44 @@ ws_gbm_buffer_new(
     return NULL;
 }
 
+static void
```
Can you add some documentation for this function and maybe even move it to the approrpiate location for static functions (end of the file)?

matthiasbeyer commented on src/compositor/buffer/gbm.c:
```diff
@@ -129,8 +162,16 @@ ws_gbm_buffer_new(
+    }
+
+    if (self->gbm_fb) {
+    }
```
Empty if there.

matthiasbeyer commented on src/compositor/cursor.c:
```diff
@@ -124,8 +126,9 @@ get_surface_under_cursor(
-        *((struct ws_surface**) target) = (struct ws_surface*) _surface;
+    if (ws_region_inside(surface->surface->input_region, x - real_x, y - real_y)) {
+        *((struct ws_abstract_shell_surface**) target) =
+                                   (struct ws_abstract_shell_surface*) _surface;
```
You don't need the second cast here, just use `surface` from above.

matthiasbeyer commented on src/compositor/cursor.c:
```diff
@@ -201,8 +204,8 @@ ws_cursor_set_position(
-    int w = ws_buffer_width((struct ws_buffer*) self->cur_mon->buffer);
-    int h = ws_buffer_height((struct ws_buffer*) self->cur_mon->buffer);
+    int w = self->cur_mon->current_mode->mode.hdisplay;
+    int h = self->cur_mon->current_mode->mode.vdisplay;
```
I'm not sure what `cur_mon` is, is it a `ws_monitor`? If yes: I think you need read locking here...

matthiasbeyer commented on src/compositor/cursor.c:
```diff
@@ -360,8 +363,11 @@ ws_cursor_get_surface_under_cursor(
+    if (!surface) {
+        return NULL;
+    }
+    return surface->surface;
```
Can be compressed to `return (surface ? surface->surface : NULL);`

neithernut replied:
Would leave this as is. Saves you from dereferencing a `NULL`-pointer on potentially dumb compilers.

matthiasbeyer commented on src/compositor/cursor.h:
```diff
@@ -57,7 +57,7 @@ struct ws_cursor {
     int x_hp; //!< @private position hotspot of the monitor
     int y_hp; //!< @private position hotspot of the monitor
-    struct ws_surface* active_surface;
+    struct ws_abstract_shell_surface* active_surface;
```
Docu missing.

matthiasbeyer commented on src/compositor/framebuffer_device.c:
```diff
@@ -137,16 +140,32 @@ ws_framebuffer_device_get_egl_display(
+    EGLint major, minor;
+
+    if (!eglInitialize(disp, &major, &minor)) {
+        return NULL;
```
I'd like to see logging here...

matthiasbeyer commented on src/compositor/module.c:
```diff
@@ -492,6 +466,15 @@ populate_connectors(void) {
+    ev_check* w,
+    int revents
+) {
+    ws_set_select(&ws_comp_ctx.monitors, NULL, NULL, ws_monitor_redraw, NULL);
```
Maybe we care about the return value here, for logging at least... would like you to add a `@todo logging` here...

matthiasbeyer commented on src/compositor/monitor.c:
```diff
@@ -172,23 +203,147 @@ ws_monitor_populate_fb(
+
+    if (eglGetError() != EGL_SUCCESS || !ret) {
+        ws_log(&log_ctx, LOG_CRIT, "Could not bind wl display to egl");
+        ws_wayland_release_display();
```
You release the display a second time here... not sure whether this is okay here...

matthiasbeyer commented on src/compositor/monitor.c:
```diff
@@ -172,23 +203,147 @@ ws_monitor_populate_fb(
+        return;
+    }
+
+    ws_wayland_release_display();
```
... and a third time...

matthiasbeyer commented on src/compositor/monitor.c:
```diff
@@ -172,23 +203,147 @@ ws_monitor_populate_fb(
+        "                     0.0, 0.0,          0.0,  1.0);"
+        "   gl_Position = vec4(position, 1.0, 1.0) * proj;"
+        "   uv = UV;"
+        "}";
```
W. T. F.?

neithernut replied:
It's a shader doing an affine transformation using a matrix.

matthiasbeyer commented on src/compositor/monitor.c:
```diff
@@ -172,23 +203,147 @@ ws_monitor_populate_fb(
+        "uniform sampler2D tex;"
+        "void main() {"
+        "   gl_FragColor = vec4(texture2D(tex, uv).rgb, 1);"
+        "}";
```
same as above.

matthiasbeyer commented on src/compositor/monitor.c:
```diff
@@ -172,23 +203,147 @@ ws_monitor_populate_fb(
+            self->current_mode->mode.vdisplay);
+
+
+    ws_monitor_redraw(self);
```
This is an aweful long ~~patch~~ diff. Please add scopes where possible.

matthiasbeyer commented on src/compositor/monitor.c:
```diff
@@ -368,3 +560,48 @@ monitor_cmp(
+    struct ws_abstract_shell_surface* surf;
+    surf = (struct ws_abstract_shell_surface*) surf_;
+
+    if (!surf->surface) {
```
Do you care about an `unlikely()` macro here?

TheNeikos replied:
It might not be that unlikely at some point

matthiasbeyer commented on src/compositor/texture.c:
```diff
@@ -71,16 +73,22 @@ int
+    // if (res < 0) {
+    //     return res;
+    // }
+    // self->obj.id = &WS_OBJECT_TYPE_ID_TEXTURE;
```
Should this stay outcommented? Maybe we can remove it?

TheNeikos replied:
Nope, this is the part which I told you about that makes it all crash...

matthiasbeyer replied:
This part? Huh? Are you sure you actually passing a `ws_texture` here? I cannot see how it could crash... maybe we need to debug a 'lil bit here! :smiley:

neithernut replied:
I, too, would like to get this investigated.

neithernut replied:
Suddenly works, according to @TheNeikos.

matthiasbeyer commented on src/compositor/wayland/abstract_shell_surface.c:
```diff
@@ -157,7 +170,20 @@ ws_abstract_shell_surface_init(
+    self->x = 1;
+    self->y = 1;
+
+    // ws_abstract_shell_surface_set_width_and_height(self, 250, 250);
```
Outcommented code fragment here...

matthiasbeyer commented on src/compositor/wayland/abstract_shell_surface.c:
```diff
@@ -254,6 +280,8 @@ shell_surface_deinit(
 
+    ws_set_remove(&shell_surf->monitor->surfaces, obj);
+
     ws_object_unref((struct ws_object*) &shell_surf->surface);
```
You could remove the cast here, too...

matthiasbeyer commented on src/compositor/wayland/abstract_shell_surface.c:
```diff
@@ -392,3 +420,24 @@ cmd_func_set_visibility(
+    shell1 = (struct ws_abstract_shell_surface*) obj1;
+    shell2 = (struct ws_abstract_shell_surface*) obj2;
+
+    struct ws_surface* s1 = (struct ws_surface*) shell1->surface;
```
Why not getting the `wl_obj` member of the surface directly here?

matthiasbeyer commented on src/compositor/wayland/abstract_shell_surface.c:
```diff
@@ -392,3 +420,24 @@ cmd_func_set_visibility(
+    struct ws_surface* s1 = (struct ws_surface*) shell1->surface;
+    struct wl_resource* r1 = ws_wayland_obj_get_wl_resource(&s1->wl_obj);
+
+    struct ws_surface* s2 = (struct ws_surface*) shell2->surface;
```
Why not getting the `wl_obj` member of the surface directly here?

matthiasbeyer commented on src/compositor/wayland/abstract_shell_surface.h:
```diff
@@ -54,6 +54,11 @@ struct ws_abstract_shell_surface {
+    uint32_t height; //!< The width of the surface
+    int32_t x; //!< The x position of the surface
+    int32_t y; //!< The y position of the surface
+    struct ws_monitor* monitor; //!< The monitor we display on
```
Visibility missing!

matthiasbeyer commented on src/compositor/wayland/buffer.c:
```diff
@@ -436,6 +474,38 @@ egl_transfer2texture(
+        return width;
+    }
+
+    return 0;
```
You can shorten this:

``` c
int32_t width = 0;
eglQ...();
return width;
```

matthiasbeyer commented on src/compositor/wayland/buffer.c:
```diff
@@ -436,6 +474,38 @@ egl_transfer2texture(
+        return height;
+    }
+
+    return 0;
```
Same as above.

matthiasbeyer commented on src/compositor/wayland/surface.c:
```diff
@@ -502,3 +486,41 @@ ws_surface_set_role(
+    glBindBuffer(GL_ARRAY_BUFFER, s->vbo);
+    glBufferData(GL_ARRAY_BUFFER, sizeof(v), v, GL_DYNAMIC_DRAW);
+
+    GLint stride = 4 * sizeof(GLfloat);
```
Care about making this a `const` ?

matthiasbeyer commented:
This is a monster of a PR. Let me suggest some things:

1) More commits would be better! I'd love to see a re-written history for this!
2) You do a lot of stuff here and I'm currently working with the abstract shell surface as well. I'd stop my work on it until your stuff is in.
3) Feel free to move files if you think they should be moved. Make sure you use `git mv` for moving/renaming!
4) More comments are warmly welcome! I do not understand most of the stuff you do here!

Awesome piece of work!

TheNeikos commented:
I am sure you saw that the name of the commit was to be reverted and the --amend commit!

I am definitely going to split it and possibly take out a few commits and submit them seperately, but I still need some refactoring, for example for OpenGL Programs etc...

Will all be done during the CCC :D

matthiasbeyer commented:
As I told you in the IRC: I was bored, that's because the comments are there... feel free to ignore them if the appropriate patch vanishes anyways! They are just suggestions for you!

31C3, we're coming! :-)

neithernut commented:
A few things which modify code on which this PR is based on are already merged. I'd suggest merging #647 first, as this PR has to be rebased anyway.

---

This PR may now be rebased.

TheNeikos committed:
Changed the egl utils to reflect correct constants

TheNeikos committed:
Added missing wayland-server include

Just including a wayland-server-protocol.h will break further includes due to
forward declarations.

TheNeikos committed:
Reimplemented gbm_surface to compositor level as it is not a buffer anymore

TheNeikos committed:
Added positioning and monitor mgmt to abs

TheNeikos committed:
Added abs to cursor and keyboard

TheNeikos committed:
Added more docmentation and egl init to fb dev

TheNeikos committed:
Added EGL Initializing to module.c

TheNeikos committed:
Rewrote texture binding

TheNeikos committed:
Added egl width/height fetching

TheNeikos committed:
Removed surface addition from compositor

TheNeikos committed:
Rewrote surface drawing

TheNeikos committed:
Added EGL drawing to monitors

TheNeikos committed:
Added ws_gl_program stubs

TheNeikos committed:
Added the correct object reference

TheNeikos committed:
Removed cmd_func_visibility

TheNeikos committed:
Removed add_surface_to_monitor compositor

TheNeikos committed:
Moved context.c to use abs

TheNeikos committed:
Removed unused format variable

neithernut commented:
Is this still WIP?

neithernut commented on src/compositor/framebuffer_device.c:
```diff
@@ -192,12 +214,26 @@ ws_framebuffer_device_get_egl_display(
             break;
         }
     }
-
```
Is there a reson to remove this line?

neithernut commented:
For book-keeping:
This PR actually targets
- #418
- #572
- #606

It also removes/obsoletes a few TODOs, so it's also related to #467.

neithernut commented:
It appears that, on my machine, the geometry of the image generated by EGL does not match the geometry of the monitor. I therefore get an distorted image. I'll try to debug this some time soon.

---

The distortion does not seem to come from a geometry mismatch, at least according to my tests. I used the following batch file (on the current `HEAD`):

```
break compositor/monitor.c:429
commands
    silent
    print "redraw"
    print self->current_mode->mode.hdisplay
    print self->current_mode->mode.vdisplay
    print self->gbm_surf->width
    print self->gbm_surf->height
    continue
end
run
```

---

By now, the line in the batch file has to be adjusted. Choose a line _after_ the `NULL`-check in `ws_monitor_redraw`.

TheNeikos committed:
Added missing docstring to cursor.h

TheNeikos committed:
Added missing ws_log to eglInitialize

TheNeikos committed:
Added documentation to egl calls in monitor.c

TheNeikos committed:
Added todo statement for regions in cursor.c

TheNeikos committed:
Added an allocation check

matthiasbeyer commented:
Output is fine, but I actually can _see_ the output, which is obviousely _not_ fine...

TheNeikos committed:
To be reverted

manuelmessner commented:
I did some debugging as well:
It seems like the problem (at least on my laptop) is the page flipping.

I could locate the failing function (which is invoked through a IOCTL) in the sources of the i915 driver of the kernel:

drivers/gpu/drm/i915/intel_display.c:9950

``` c
/*
 * TILEOFF/LINOFF registers can't be changed via MI display flips.
 * Note that pitch changes could also affect these register.
 */
if (INTEL_INFO(dev)->gen > 3 &&
    (fb->offsets[0] != crtc->primary->fb->offsets[0] ||
     fb->pitches[0] != crtc->primary->fb->pitches[0]))
   return -EINVAL;
```

The results of the single comparisons are the following:

``` c
INTEL_INFO(dev)->gen == 7
fb->offsets[0] != crtc->primary->fb->offsets[0] == 0
fb->pitches[0] != crtc->primary->fb->pitches[0] == 1
```

TheNeikos committed:
To be reverted

TheNeikos commented:
It's been a good run o7
